### ISSUE TITLE
Synchronous JS interop: casting to IJSInProcessObjectReference throws

### ISSUE THREAD

**Issue by ChristianWeyer** [dotnet/aspnetcore#26588]
Looking at the preliminary docs here:
https://github.com/dotnet/AspNetCore.Docs/blob/ddc899a36770fca83cff263244df7523cea2aeb6/aspnetcore/blazor/webassembly-performance-best-practices.md

Using .NET 5: 5.0.0-rc.2.20474.6

In my Blaror WASM app I have a service like this:

```
public class WebcamService : IWebcamService, IAsyncDisposable
    {
        private readonly Lazy<Task<IJSObjectReference>> _moduleTask;

        public WebcamService(IJSRuntime jsRuntime)
        {
            _moduleTask = new(() => jsRuntime.InvokeAsync<IJSObjectReference>(
               "import", "./jsinterop/webcam.js").AsTask());
        }

        public async Task StartVideo(WebcamOptions options)
        {
            var module = await _moduleTask.Value;
            ((IJSInProcessObjectReference)module).InvokeVoid("startVideo", options);
        }

        public async Task TakePicture()
        {
            var module = await _moduleTask.Value;
            ((IJSInProcessObjectReference)module).InvokeVoid("takePicture");
        }

        public async ValueTask DisposeAsync()
        {
            if (_moduleTask.IsValueCreated)
            {
                var module = await _moduleTask.Value;
                await module.DisposeAsync();
            }
        }
    }
```

When calling e.g. the StartVideo method, the code throws:

```
blazor.webassembly.js:1 crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Specified cast is not valid.
System.InvalidCastException: Specified cast is not valid.
   at ConfTool.Client.Webcam.WebcamService.StartVideo(WebcamOptions options) in C:\sources\blazor-webassembly-demo\Client\Webcam\WebcamService.cs:line 20
   at ConfTool.Client.Webcam.Webcam.StartVideoAsync() in C:\sources\blazor-webassembly-demo\Client\Webcam\Webcam.razor:line 40
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle)
```

Not sure what could be wrong here.

Repro can be found here:
https://github.com/thinktecture/blazor-webassembly-demo/blob/net5-sync-js/Client/Webcam/WebcamService.cs#L20

Thanks!

**Comment by smartprogrammer93** [comment]
@ChristianWeyer can you put the following before the casting and tell us the output
`Console.WriteLine(module.GetType().FullName);`

**Comment by ChristianWeyer** [comment]
Sure.
Output is:

`Microsoft.JSInterop.Implementation.JSObjectReference`

**Comment by smartprogrammer93** [comment]
I Believe the type should be WebAssemblyJSObjectReference instead. I checked the WebassemblyJSRuntime code and it does return that. so i believe the problem is that the injected IJSRuntime is not the correct one. @SteveSandersonMS  could it be because his project is Asp.net core hosted?!!

**Comment by pranavkm** [comment]
Our implementation produces different flavors of IJSObjectReference depending on how it's retrieved. Reading it as a `IJSObjectReference` vs `IJSInProcessObjectReference` vs `IJSUnmarshalledJSObjectReference` produces slightly different results which might makes casting between them problematic.

For your scenario, I would recommend reading the value as `IJSUnmarshalledObjectReference`.  

@SteveSandersonMS there's a fairly concise change we can make - https://github.com/dotnet/aspnetcore/pull/27524/files - to able to rectify the behavior. If you feel strongly about it, we should try and bring it thru tactics to fix in 5.0,1.

**Comment by EnsignPayton** [comment]
I'm able to work around this like so, but it would be nice if reflection wasn't needed here.

```cs
private readonly IJSRuntime _jsRuntime;

...

private IJSUnmarshalledObjectReference GetUnmarshalledModule(IJSObjectReference module)
{
    if (module is IJSUnmarshalledObjectReference uModule) return uModule;

    var wasm = typeof(WebAssemblyJSRuntime).Assembly;

    var moduleId = module.GetType()
        .GetProperty("Id", BindingFlags.Instance | BindingFlags.NonPublic)
        .GetValue(module);

    return wasm.GetType("Microsoft.JSInterop.WebAssembly.WebAssemblyJSObjectReference")
        .GetConstructor(new[] { typeof(WebAssemblyJSRuntime), typeof(long) })
        .Invoke(new[] { _jsRuntime, moduleId }) as IJSUnmarshalledObjectReference;
}
```
